design/src/rfcs/rfc0003_presigning_api.md

```diff
@@ -0,0 +1,192 @@
+- Valid request input
+- Endpoint
+- Credentials to sign with
+- Signing implementation
+
+The AWS middleware provides everything except the request, and the request is provided as part
+of the fluent builder API. The generated code needs to be able to run the middleware to fully populate
+a request property bag, but not actually dispatch it. Additionally, the signing headers added by the signing
+implementation will need to be moved into query parameters.
+
+Today, request dispatch looks as follows:
+1. The customer creates a new fluent builder by calling `client.operation_name()`, fills in inputs, and then calls `send()`.
+2. `send()`:
+   1. Builds the final input struct, and then calls its `make_operation()` method with the stored config to create a Smithy `Operation`.
+   2. Calls the underlying Smithy client with the operation.
+3. The Smithy client constructs a Tower Service with AWS middleware and a dispatcher at the bottom, and then executes it.
+4. The middleware acquire and add required signing parameters (region, credentials, endpoint, etc) to the request property bag.
+5. The SigV4 signing middleware signs the request by adding HTTP headers to it.
+6. The dispatcher makes the actual HTTP request and returns the response all the way back up the Tower.
+
+Pre-signing will take advantage of a lot of these same steps, but will cut out the `Operation` and
+replace the dispatcher with a pre-signed URL generator:
+1. The customer creates a new fluent builder by calling `client.operation_name()`, fills in inputs, and then calls `presigned()`.
+2. `presigned()`:
+   1. Builds the final input struct, and then calls a new `make_request()` method with the stored config.
+   2. Constructs a Tower Service with `AwsMiddleware` layered in, and a `PresignedUrlGeneratorLayer` at the bottom.
+   3. Calls the Tower Service and returns its result
+3. The `AwsMiddleware` will sign the request by adding signature headers to it.
+4. `PresignedUrlGeneratorLayer` would take the requests URI, and append signature headers to it as query parameters. Then return the result as a string.
+
+It should be noted that the `presigned()` function above is on the generated input struct, so implementing this for
+the input API is identical to implementing it for the fluent client.
+
+All the code for the new `make_request()` is already in the existing `make_operation()` and will just need to be split out.
+
+### Modeling Pre-signing
+
+AWS models don't currently have any information about which operations can be pre-signed.
+To work around this, the Rust SDK will create a synthetic trait to model pre-signing with, and
+apply this trait to known pre-signed operations via customization. The code generator will
+look for this synthetic trait when creating the fluent builders and inputs to know if a
+`presigned()` method should be added.
+
+Unresolved Questions
+--------------------
+
+- Where should `PresigningConfig` live?
```

Comment:
- if we are confident in global setting stability, in `aws-sig-auth`.
- if we think services might end up with their own settings, inlined into individual services